url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+    	JSONObject message;
+    	InetSocketAddress address;
+        try{
+        switch(messageType){
+            case JOINING_NETWORK_SIMPLIFIED:
+            	//update routing table with single contact
+            	singleUpdateRouteTable(node_id,packet.getAddress().getHostAddress());
+            	if(nodeID == target_id){
+            		message = getRoutingInfoJson(jsonObj,nodeID);
+            		address = new InetSocketAddress(packet.getAddress(),NetworkProperties.nPort);
+            		sendQueue.put(jsonToDatagram(message,address));
+            	}else{
+            		closestNode = getClosestNode(target_id);            		
+            		if(closestNode != null){// maybe I am closest but still route message to at least one other node
+                		message = new JSONObject();
+                		message.put("type", "JOINING_NETWORK_RELAY_SIMPLIFIED");
+                		message.put("node_id",String.valueOf(node_id));
+                		message.put("target_id", String.valueOf(closestNode));
+                		message.put("gateway_id", String.valueOf(nodeID));
+                		address = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort);
+                		sendQueue.put(jsonToDatagram(message,address));	
+            		}else{ // not target and does not have anything in the routing table
+            			
+            		}
+            	}
+                break;
+            case JOINING_NETWORK_RELAY_SIMPLIFIED:
+            	singleUpdateRouteTable(node_id,packet.getAddress().getHostAddress());// I assume the gateway is sending this message even if it actually relayed(for now)
+            	if(target_id == nodeID){
+	            	closestNode = getClosestNode(node_id);
+	            	//if there is nothing in the routing table
+	            	if (closestNode == null){
+	                   	message = getRoutingInfoJson(jsonObj,target_id);
+	                	address = new InetSocketAddress(packet.getAddress(),packet.getPort());//this wouldn't have effect in simulation    		
+	            	}else{//else forward message. No need to reconstruct the whole message
+	            		message = jsonObj;
+	            		message.put("target_id", String.valueOf(closestNode.getId()));
+	            		address = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort);	            		
+	            	}
+	            	sendQueue.put(jsonToDatagram(message,address));
+            	}
+                break;
+                
+            case ROUTING_INFO:
+            	singleUpdateRouteTable(sender_id,ip_address);
+            	updateRouteTable((JSONArray)jsonObj.get("route_table"));//dump the all routing_info messages in the routing table
+        		message = jsonObj;
+            	if(gateway_id == nodeID){ // if current is gateway
+                    assert  routingTable.get(node_id).getIp() == "127.0.0.1" : "ip address of node in routing info is wrong";//TODO remember to disable assertions for non-testing environment
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.